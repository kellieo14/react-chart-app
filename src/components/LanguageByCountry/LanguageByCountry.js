import React, {useState, useEffect} from 'react';
import HorizontalChart from '../Graphs/HorizontalBar/HorizontalBar';
import {sort} from '../../Utils';

export const extractLanguages = (countries, languages) => {
    for (let country in countries) {
        countries[country].languages.forEach(element => {
            languages.push(element.name);
        })
        languages.sort()
    }
}

export const getLanguageCounts = (result, languages) => {
    for (let language in languages) {
        if (result[languages[language]] === undefined) {
            result[languages[language]] = 0;
        } 
        result[languages[language]]++;
    }
}

export const  sortLanguagesByPopularity = (result, languageCountArr, labels, countryCount) => {
    for (let language in result) {
        languageCountArr.push([language, result[language]]);
    }
    sort(languageCountArr, 'language', 8, labels, countryCount)
}

export const setTopLanguagesByCountry = (labels, setData, countryCount) => {
    if (labels.length > 0) {
        setData({
            title: 'Most Common Languages by Country Count',
            labels: labels,
            datasets: [
            {
                label: 'Countries',
                backgroundColor: 'rgba(255,99,132,0.1)',
                borderColor: 'rgba(255,99,132,1)',
                borderWidth: 1,
                hoverBackgroundColor: 'rgba(255,99,132,0.4)',
                hoverBorderColor: 'rgba(255,99,132,1)',
                data: countryCount
            }
            ]
        });
    } 
}

const LanguageByCountry = (props) => {
    const countries = props.countries;
    const [data, setData] = useState({});

    useEffect(() => {
        let languages = [];
        let labels = [];
        let countryCount = [];
        let result={};
        let languageCountArr = [];

        extractLanguages(countries, languages);
        getLanguageCounts(result, languages);
        sortLanguagesByPopularity(result, languageCountArr, labels, countryCount);
        setTopLanguagesByCountry(labels, setData, countryCount);
    }, [countries]);

    return (
        <div>
            <HorizontalChart data={data}/>
        </div>
      );
}

export default LanguageByCountry;


