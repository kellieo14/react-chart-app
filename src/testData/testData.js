export const countries = [
    {
        name: 'China', 
        region: 'Asia', 
        languages: [
            { iso639_1: "zh", iso639_2: "zho", name: "Chinese", nativeName: "中文 (Zhōngwén)"}
        ], 
        population: 1377422166,
        area: 9640011,
    }, 
    {
        name: 'India', 
        region: 'Asia',
        languages: [
            {iso639_1: "hi", iso639_2: "hin", name: "Hindi", nativeName: "हिन्दी"},
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 1295210000,
        area: 3287590,
    }, 
    {
        name: 'United States of America', 
        region: 'Americas', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 323947000,
        area: 9629091,
    },
    {
        name: 'Nigeria', 
        region: 'Africa', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 186988000,
        area: 923768,
    }, 
    {
        name: 'France', 
        region: 'Europe', 
        languages: [
            {iso639_1: "fr", iso639_2: "fra", name: "French", nativeName: "français"}
        ], 
        population: 66710000,
        area: 640679,
    }, 
    {
        name: 'Australia', 
        region: 'Oceania', 
        languages: [
            {iso639_1: "en", iso639_2: "eng", name: "English", nativeName: "English"}
        ], 
        population: 24117360,
        area: 7692024,
    }, {
        name: 'Polarland', 
        region: 'Polar', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 11239004,
        area: 1284000,
    }, {
        name: 'Otherland', 
        region: 'Other', 
        languages: [
            {iso639_1: "es", iso639_2: "spa", name: "Spanish", nativeName: "Español"}
        ], 
        population: 100,
        area: 619745,
    }
];