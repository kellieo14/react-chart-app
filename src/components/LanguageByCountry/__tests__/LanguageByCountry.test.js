import React from 'react';
import * as reactHooks from 'react';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TestRenderer from 'react-test-renderer';
import { countries } from '../../../testData/testData';
import LanguageByCountry, { extractLanguages, getLanguageCounts, sortLanguagesByPopularity } from '../LanguageByCountry';
Enzyme.configure({adapter: new Adapter()});

describe('LanguageByCountry', () => {

    let spyUseState;
    let spyUseEffect;
    let mockSetData;

    beforeEach(() => {
        mockSetData = jest.fn();
        spyUseState = jest.spyOn(reactHooks, 'useState')
            .mockImplementationOnce(() => [{}, mockSetData]);
        spyUseEffect = jest.spyOn(reactHooks, 'useEffect');
    })
    const expectedLanguages = [
        'Chinese', 
        'English', 
        'English', 
        'English',
        'English', 
        'French', 
        'Hindi',
        'Spanish',
        'Spanish',
    ];

    const expectedLanguageCounts = {
        Chinese: 1, 
        English: 4, 
        French: 1, 
        Hindi: 1, 
        Spanish: 2,
    }

    it('should match snapshot', () => {
        const testRenderer = TestRenderer.create(<LanguageByCountry />);
        expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    it('should call useSatate and useEffect', () => {
        mount(<LanguageByCountry countries={countries} />);
        expect(spyUseState).toHaveBeenCalled();
        expect(spyUseEffect).toHaveBeenCalled();
        expect(mockSetData).toHaveBeenCalled();
    });

    it('should extract languages from countries', () => {
        let languages = [];
        extractLanguages(countries, languages);
        expect(languages).toEqual(expectedLanguages);
    });

    it('should count the occurence of each language', () => {
        let result = {};
        getLanguageCounts(result, expectedLanguages);
        expect(result).toEqual(expectedLanguageCounts);
    });

    it('should sort language by popularity', () => {
        let languageCountArr = [];
        let labels = [];
        let countryCount = [];

        const expectedLanguageCountArr = [
            ['English', 4], 
            ['Spanish', 2], 
            ['Chinese', 1], 
            ['French', 1], 
            ['Hindi', 1]
        ];

        const expectedLabels = ['English', 'Spanish', 'Chinese', 'French', 'Hindi'];

        const expectedCountryCount = [4, 2, 1, 1, 1];

        sortLanguagesByPopularity(expectedLanguageCounts, languageCountArr, labels, countryCount);
        expect(languageCountArr).toEqual(expectedLanguageCountArr);
        expect(labels).toEqual(expectedLabels);
        expect(countryCount).toEqual(expectedCountryCount);
    });
})