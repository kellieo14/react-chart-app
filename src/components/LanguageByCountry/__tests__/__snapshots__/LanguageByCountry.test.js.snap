// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageByCountry should match snapshot 1`] = `
<div>
  <div
    className="horizontalBar"
  />
</div>
`;
